### PR TITLE
Add comments when dumping

### PR DESCRIPTION
**Pull request by solaluset**
I thought it'd be useful to have a possibility to add comments when dumping by passing `comments=...` to `dump(s)`. Also I've added `trailing_comma` arg that adds trailing commas everywhere.

### CODE REVIEW

**Comment by n-takumasa** [comment]
Thanks for the suggestion.

What is your intended use-case?
Considering symmetry, `load` should add a function to get comments.
But I think comments are not data.

**Comment by solaluset** [comment]
Exactly, comments are not data. That's why I think it shouldn't be possible to extract them. But adding them is a different thing. My program generates a config file and I want the user to see what is the meaning of some settings. You can see the code here:

https://github.com/Krutyi-4el/DandelionMusic/blob/44e34d1ce23709f6214a6d5ea3d5ff402d27078c/config/config.py#L176

File: jsonc/__init__.py [hunk 1]
```diff
@@ -60,6 +81,38 @@ def _remove_trailing_comma(text: str) -> str:
...
+    return "\n".join(
+        " " * indent + "// " + line if line else "" for line in text.splitlines()
+    )
+
+
+def _get_comments(comments, key):
+    if isinstance(comments, dict):
+        if isinstance(key, str):
+            key = json.loads(key)
+        comments = comments.get(key)
```
Comment (n-takumasa) [inline jsonc/__init__.py]: may be None

File: jsonc/__init__.py [hunk 2]
```diff
@@ -60,6 +81,38 @@ def _remove_trailing_comma(text: str) -> str:
...
+    )
+
+
+def _make_comment(text: str, indent=0) -> str:
+    return "\n".join(
+        " " * indent + "// " + line if line else "" for line in text.splitlines()
+    )
+
+
+def _get_comments(comments, key):
```
Comment (n-takumasa) [inline jsonc/__init__.py]: plz add annotations

Comment (n-takumasa) [reply jsonc/__init__.py]: tuple[str | None, Comments | None] ?

**Review by n-takumasa — CHANGES_REQUESTED** [review]
request changes:
* add annotations for `_get_comments`
* add error handling when any inconsistencies occur with data and comments

**Comment by n-takumasa** [comment]
I see your intent, I thin it's fine for that use-case.
However, I have made some requests to improve robustness, please consider them.

File: jsonc/__init__.py (hunk 1 above)
Comment (solaluset) [reply jsonc/__init__.py]: It's intentional, not all elements have to have comments

**Comment by solaluset** [comment]
> add error handling when any inconsistencies occur with data and comments

Not sure about this one. Should the script validate that all keys in dictionary are strs when commenting dict and ints when list?

**Comment by n-takumasa** [comment]
> > add error handling when any inconsistencies occur with data and comments
> 
> Not sure about this one. Should the script validate that all keys in dictionary are strs when commenting dict and ints when list?

I think that at least occurence a warning will help.

```py
print(jsonc.dumps({"b": 1}, indent=2, comments={"a": "foo"}))
# {
#   "b": 1
# }
```

Also, when you give `comments` arg with no `indent` arg, a warning should be occur.

**Comment by solaluset** [comment]
I've added warns for unused comment keys. However, it has a side effect of removing all used ones from original dictionary. I'll add `copy.deepcopy()` if you think it may become an issue.

**Comment by n-takumasa** [comment]
Nice work

`deepcopy` is necessary because I usually think that functions must not have any side effect on their arguments.

**Review by n-takumasa — APPROVED** [review]

**Comment by n-takumasa** [comment]
merged. thanks!

**Pull request merged**